docs/src/main/sphinx/sql/deny.rst

```diff
@@ -0,0 +1,51 @@
+====
+DENY
+====
+
+Synopsis
+--------
+
+.. code-block:: text
+
+    DENY ( privilege [, ...] | ( ALL PRIVILEGES ) )
+    ON ( table_name | TABLE table_name | SCHEMA schema_name)
+    TO ( user | USER user | ROLE role )
+
+Description
+-----------
+
+Denies the specified privileges to the specified grantee.
+
+Specifying ``ALL PRIVILEGES`` denies :doc:`delete`, :doc:`insert`, :doc:`update` and :doc:`select` privileges.
+
+Specifying ``ROLE PUBLIC`` denies privileges to the ``PUBLIC`` role and hence to all users.
+
+Deny on a table grants the specified privilege on all current and future columns of the table.
+
+Deny on a schema denies the specified privilege on all current and future columns of all current and future tables of the schema.
+
```

Comment:
There is (at the moment) no rule in place as far I see in the code how `DENY` interacts with `GRANT` statement. Probably this is due to the fact that `DENY` has not yet been implemented in any connector.

Reply:
- `DENY` overrides all `GRANT`s “covered” by it, so if you `GRANT` privileges to a table and then `DENY` access to a schema containing that table, the table grant will be removed. That way if you later `REVOKE` the `DENY`, the access won’t come back by itself - it’ll have to be re-added explicitly.
- `GRANT` will remove a `DENY` if it is at the exact “coordinates”. So for example you `DENY` access to a schema, then `GRANT` to the same schema, it will replace the `DENY`.
- `REVOKE` will remove any `GRANT` or `DENY`.


This should probably land in https://trino.io/docs/current/language/sql-support.html#security-operations (in case we go further with this PR).

Courtesy of @ksobolew

Reply:
Yes, there are no implementations in Trino as of yet, but this is the semantics we agreed on in case someone wants to actually implement it.

Reply:
Agreed. But nothing prevents anyone from implementing whatever they want out of tree. i.e. documenting the expected behaviour when the actual behaviour is no-op doesn't make a lot of sense to me. It'll just feed confusion where people will try out DENY statements and get errors since no access control implements it today.